engineapi/engineapi/middleware.py

```diff
@@ -199,3 +212,157 @@ async def dispatch(
         request.state.token = user_token
 
         return await call_next(request)
+
+
+def parse_origins_from_resources(origins: List[str]) -> Set[str]:
+    """
+    Parse list of CORS origins with HTTP validation and remove duplications.
+    """
+    resource_origins_set = set()
+    for resource in origins:
+        origins = resource.resource_data.get("origins", [])
```

Comment:
I really don't understand what this line of code is doing.

`origins` is the input variable, of type `List[str]`. `resource` is an item from `origins`, which means it is a string. What is `resource.resource_data`, then?

And why are you modifying the `origins` argument to the function inside the loop?

Reply:
This resource is where we store user provided origins. So it could have multiple of them. I decided to store this as one resource. Of course we can create for each origin new resource.

Reply:
Remove modification of `origins`